File: python/430. Flatten a Multilevel Doubly Linked List.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""
class Solution(object):
    def flatten(self, head):
        """
        :type head: Node
        :rtype: Node
        """
        st = []
        pre = None
        p = head
        while p:
            pre = p
            if p.child:
                if p.next:
                    st.append(p.next)
                p.child.prev = p
                p.next= p.child
                p.child = None
                p = p.next
            else:
                p = p.next
            
            if p is None:
                if st:
                    p = st.pop()
                    pre.next = p
                    p.prev = pre
        return head

# 执行用时 :
# 32 ms
# , 在所有 Python 提交中击败了
# 92.77%
# 的用户
# 内存消耗 :
# 12.3 MB
# , 在所有 Python 提交中击败了
# 97.01%
# 的用户

File: go/src/leetcode/430. Flatten a Multilevel Doubly Linked List/430. Flatten a Multilevel Doubly Linked List.go
package main

import "fmt"

type Node struct {
	Val int
	Prev *Node
	Next *Node
	Child *Node
}

func flatten(root *Node) *Node {
	if root == nil {
		return nil
	}
	if root.Child != nil {
		sub :=  flatten(root.Child)
		p := sub
		for p.Next != nil {
			p = p.Next
		}
		p.Next = root.Next
		if root.Next != nil {
			root.Next.Prev = p
		}

		root.Next = sub
		sub.Prev =  root

		root.Child = nil
	}
	root.Next = flatten(root.Next)
	return root
}

func main() {
	node1 := &Node{Val: 1}
	node2 := &Node{Val: 2}
	node3 := &Node{Val: 3}
	node1.Child = node3
	node1.Next = node2
	node2.Prev = node1

	root := flatten(node1)
	fmt.Println(root)
}
